File: frontend/src/api/member.ts
import {
  CreateMemberPayload,
  CreateMemberResponse,
  DeleteMemberPayload,
  DeleteMemberResponse,
  UpdateMemberPayload,
  UpdateMemberResponse
} from 'generated/models'
import { RequestAPI } from './request'

// All endpoints under the '/members' prefix
export class MemberAPI {
  private req: RequestAPI

  constructor() {
    this.req = new RequestAPI('/members')
  }

  async createMember(
    payload: CreateMemberPayload
  ): Promise<CreateMemberResponse> {
    return this.req.post('/', payload)
  }

  async editMember(
    payload: UpdateMemberPayload
  ): Promise<UpdateMemberResponse> {
    return this.req.put('/', payload)
  }

  async deleteMember(
    payload: DeleteMemberPayload
  ): Promise<DeleteMemberResponse> {
    return this.req.delete(`/${payload.id}`)
  }
}

File: backend/controllers/member.go
package controllers

import (
	"fmt"
	"net/http"

	"github.com/EmilyOng/cvwo/backend/models"
	memberService "github.com/EmilyOng/cvwo/backend/services/member"
	errorUtils "github.com/EmilyOng/cvwo/backend/utils/error"

	"github.com/gin-gonic/gin"
)

func CreateMember(c *gin.Context) {
	var payload models.CreateMemberPayload

	err := c.ShouldBindJSON(&payload)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, models.Response{Error: error_UNEXPECTED})
		return
	}

	createMemberResponse := memberService.CreateMember(payload)
	c.JSON(errorUtils.MakeResponseCode(createMemberResponse.Response), createMemberResponse)
}

func UpdateMember(c *gin.Context) {
	var payload models.UpdateMemberPayload

	err := c.ShouldBindJSON(&payload)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, models.Response{Error: error_UNEXPECTED})
		return
	}

	updateMemberResponse := memberService.UpdateMember(payload)
	c.JSON(errorUtils.MakeResponseCode(updateMemberResponse.Response), updateMemberResponse)
}

func DeleteMember(c *gin.Context) {
	var memberID uint8
	fmt.Sscan(c.Param("member_id"), &memberID)

	deleteMemberResponse := memberService.DeleteMember(models.DeleteMemberPayload{ID: memberID})
	c.JSON(errorUtils.MakeResponseCode(deleteMemberResponse.Response), deleteMemberResponse)
}
